Source/ARTRealtimePresence.m

```diff
@@ -780,85 +778,110 @@ - (void)reenterInternalMembers {
+            }
+            break;
+        default:
+            break;
+    }
 
-- (void)internalAdd:(ARTPresenceMessage *)message {
-    [self internalAdd:message withSessionId:_syncSessionId];
+    if (memberUpdated) {
+        [self broadcast:message];
+    }
+    else {
+        ARTLogDebug(_logger, @"Presence member \"%@\" with action %@ has been ignored", message.memberKey, ARTPresenceActionToStr(message.action));
+    }
+    [_beforeSyncMembers removeObjectForKey:message.memberKey]; // RTP19
 }
 
-- (void)internalAdd:(ARTPresenceMessage *)message withSessionId:(NSUInteger)sessionId {
-    message.syncSessionId = sessionId;
-    [_members setObject:message forKey:message.memberKey];
-    // Local member
-    if ([message.connectionId isEqualToString:self.connectionId]) {
-        _internalMembers[message.clientId] = message;
-        ARTLogDebug(_logger, @"local member %@ with action %@ has been added", message.memberKey, ARTPresenceActionToStr(message.action).uppercaseString);
+- (BOOL)addMember:(ARTPresenceMessage *)message {
+    ARTPresenceMessage *existing = [_members objectForKey:message.memberKey];
+    if (existing && [existing isNewerThan:message]) {
+        return false;
     }
+    _members[message.memberKey] = message;
+    return true;
 }
 
-- (void)internalRemove:(ARTPresenceMessage *)message {
-    [self internalRemove:message force:false];
+- (BOOL)addInternalMember:(ARTPresenceMessage *)message {
+    ARTPresenceMessage *existing = [_internalMembers objectForKey:message.clientId];
+    if (existing && [existing isNewerThan:message]) {
+        return false;
+    }
+    _internalMembers[message.clientId] = message;
+    ARTLogDebug(_logger, @"local member %@ with action %@ has been added", message.clientId, ARTPresenceActionToStr(message.action).uppercaseString);
+    return true;
 }
 
-- (void)internalRemove:(ARTPresenceMessage *)message force:(BOOL)force {
-    if ([message.connectionId isEqualToString:self.connectionId] && !message.isSynthesized) {
-        [_internalMembers removeObjectForKey:message.clientId];
+- (BOOL)removeMember:(ARTPresenceMessage *)message {
+    ARTPresenceMessage *existing = [_members objectForKey:message.memberKey];
+    if (existing && [existing isNewerThan:message]) {
+        return false;
     }
+    [_members removeObjectForKey:message.memberKey];
+    return existing.action != ARTPresenceAbsent;
+}
 
-    const BOOL syncInProgress = self.syncInProgress;
-    if (!force && syncInProgress) {
-        ARTLogDebug(_logger, @"%p \"%@\" should be removed after sync ends (syncInProgress=%d)", self, message.clientId, syncInProgress);
-        message.action = ARTPresenceAbsent;
-        // Should be removed after Sync ends
-        [self internalAdd:message withSessionId:message.syncSessionId];
-    }
-    else {
-        [_members removeObjectForKey:message.memberKey];
+- (BOOL)removeInternalMember:(ARTPresenceMessage *)message {
```

Comment:
No need to return values here, since they are not being used right?

Reply:
You use the return value [here](https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/realtime_presence.go#L323).

Reply:
It's used the same way in this PR.

Reply:
My bad, removed - https://github.com/ably/ably-cocoa/pull/1891/commits/7ae70b3a8efd93ae93643657d7de0c2abdf4a2a1